autograd/core.py

```diff
@@ -52,24 +66,40 @@ class primitive(object):
     def __init__(self, fun):
         self.fun = fun
         self.vjps = {}
+        self.jvps = {}
         self.zero_vjps = set()
         self.__name__ = fun.__name__
         self.__doc__ = fun.__doc__
 
     def __call__(self, *args, **kwargs):
         argvals = list(args)
-        progenitors = set()
+        progenitors = dict()
         parents = []
         for argnum, arg in enumerate(args):
             if isnode(arg):
                 argvals[argnum] = arg.value
                 if argnum in self.zero_vjps: continue
                 parents.append((argnum, arg))
-                progenitors.update(arg.progenitors & active_progenitors)
+                progenitors.update({progenitor: val for progenitor, val in arg.progenitors.items()
+                                   if progenitor in active_progenitors})
 
         result_value = self.fun(*argvals, **kwargs)
         if progenitors:
-            return new_node(result_value, (self, args, kwargs, parents), progenitors)
+            result = new_node(result_value, (self, args, kwargs, parents), progenitors)
+            ingrads = defaultdict(list)
+            for argnum, arg in parents:
+                for progenitor, forward_grad in arg.progenitors.items():
+                    if forward_grad is None or progenitor not in active_progenitors:
+                        continue
+                    active_progenitors.remove(progenitor)
+                    ingrad = self.jvp(argnum, forward_grad, result, arg.vspace,
+                                      result.vspace, args, kwargs)
+                    active_progenitors.add(progenitor)
+                    ingrads[progenitor].append(ingrad)
+                    assert_vspace_match(ingrad, result.vspace, self)
+            result.progenitors.update({progenitor: vsum(result.vspace, *ingrads[progenitor])
+                                       for progenitor in ingrads})
```

Comment:
↑ This block should maybe be moved into a separate `update_forward_grads` function